parent spec artifacts on spec revisions

theganyo opened the pull request:
Partially addresses #774 by moving artifacts from Specs to Spec Revisions. Deployment Revisions will be addressed similarly in a following PR.

timburks commented:
I thought it would be useful to exercise this by bulk-uploading discovery specs twice (to get two revisions of every spec) and then attaching some artifacts, but while running the bulk uploader the second time (with a postgres backend), I got a flood of these errors:
```
 INFO[0293] Handling request.         method=UpdateApiVersion request_id=25d73b36 uid=91a27cd8
 ERROR[0293] Aborted.                  duration=1.554354ms error=rpc error: code = AlreadyExists desc = pq: duplicate key value violates unique constraint "versions_pkey" method=UpdateApiVersion request_id=25d73b36 status_code=Aborted uid=91a27cd8
  INFO[0293] Handling request.         method=UpdateApiVersion request_id=deaa6df6 uid=91a27cd8
 ERROR[0293] Aborted.                  duration=1.610005ms error=rpc error: code = AlreadyExists desc = pq: duplicate key value violates unique constraint "versions_pkey" method=UpdateApiVersion request_id=deaa6df6 status_code=Aborted uid=91a27cd8
  INFO[0296] Handling request.         method=UpdateApiVersion request_id=705adf3f uid=91a27cd8
 ERROR[0296] Aborted.                  duration=1.523023ms error=rpc error: code = AlreadyExists desc = pq: duplicate key value violates unique constraint "versions_pkey" method=UpdateApiVersion request_id=705adf3f status_code=Aborted uid=91a27cd8
  INFO[0296] Handling request.         method=UpdateApiVersion request_id=ec09fc4c uid=91a27cd8
 ERROR[0296] Aborted.                  duration=1.598264ms error=rpc error: code = AlreadyExists desc = pq: duplicate key value violates unique constraint "versions_pkey" method=UpdateApiVersion request_id=ec09fc4c status_code=Aborted uid=91a27cd8
  INFO[0296] Handling request.         method=UpdateApiVersion request_id=1189ee48 uid=91a27cd8
 ERROR[0296] Aborted.                  duration=1.520873ms error=rpc error: code = AlreadyExists desc = pq: duplicate key value violates unique constraint "versions_pkey" method=UpdateApiVersion request_id=1189ee48 status_code=Aborted uid=91a27cd8
  INFO[0296] Handling request.         method=UpdateApiVersion request_id=1fc28347 uid=91a27cd8
 ERROR[0296] Aborted.                  duration=1.553934ms error=rpc error: code = AlreadyExists desc = pq: duplicate key value violates unique constraint "versions_pkey" method=UpdateApiVersion request_id=1fc28347 status_code=Aborted uid=91a27cd8
  INFO[0297] Handling request.         method=UpdateApiVersion request_id=8376b607 uid=91a27cd8
 ERROR[0297] Aborted.                  duration=1.522513ms error=rpc error: code = AlreadyExists desc = pq: duplicate key value violates unique constraint "versions_pkey" method=UpdateApiVersion request_id=8376b607 status_code=Aborted uid=91a27cd8
  INFO[0297] Handling request.         method=UpdateApiVersion request_id=38814ce5 uid=91a27cd8
 ERROR[0297] Aborted.                  duration=1.646625ms error=rpc error: code = AlreadyExists desc = pq: duplicate key value violates unique constraint "versions_pkey" method=UpdateApiVersion request_id=38814ce5 status_code=Aborted uid=91a27cd8
```

shrutiparabgoogle commented on cmd/registry/controller/controller_test.go:
```diff
@@ -774,6 +778,30 @@ func TestMultipleEntitiesArtifacts(t *testing.T) {
+						ApiID:      a.ApiID(),
+						VersionID:  a.VersionID(),
+						SpecID:     a.SpecID(),
+						RevisionID: a.RevisionID(),
```
This flow is a little confusing. We parse the `GeneratedResource` (which doesn't have a revisionID) into a `SpecRevision`, fetch the `SpecRevision` and use that revisionID to populate the `GeneratedResource`.

Specially the `RevisionID: a.RevisionID()` is quite confusing. Can we instead parse the `GeneratedResource` into a `Spec` -> fetch that spec and then use the spec's revision ID to populate the `GeneratedResource`. This type of flow will avoid the confusion I think.

theganyo replied:
Sounds good.

shrutiparabgoogle commented on cmd/registry/patterns/resources.go:
```diff
@@ -52,6 +54,10 @@ func (s SpecName) Spec() string {
 	return s.Name.String()
 }
 
+func (s SpecName) Revision() string {
```
This method can be unexported because it won't be accessible from the `ResourceName` interface.

theganyo replied:
ok

shrutiparabgoogle commented on cmd/registry/patterns/resources.go:
```diff
@@ -321,6 +331,10 @@ func (ar ArtifactName) ParentName() ResourceName {
 			Name: spec,
 		}
+	} else if rev, err := names.ParseSpecRevision(parent); err == nil {
+		return SpecName{
+			Name: rev.Spec(),
+		}
```
Shouldn't this be:

```
return SpecName{
			Name: rev.Spec(),
                         RevisionID: rev.RevisionID,
		}

```


Should we also support revision collections in the previous block for matching collections?
Example: `projects/demo/locations/global/apis/-/versions/-/specs/-@-/artifacts/-` 
Not sure if the names package supports this yet.

theganyo replied:
Looking into this... Adding RevisionID to this breaks [this controller test](https://github.com/apigee/registry/blob/1777bc258f5719f80dd4cb9d3b8dd4fa285c78de/cmd/registry/controller/controller_test.go#L659) - it generates double the actions: a set of update actions for SpecRevisions and a set of create actions for Specs. So, I'm working to understand that.

theganyo commented:
> I thought it would be useful to exercise this by bulk-uploading discovery specs twice (to get two revisions of every spec) and then attaching some artifacts, but while running the bulk uploader the second time (with a postgres backend), I got a flood of these errors

Are you saying this is a new issue? Or just an observation you happened to perform on this PR?

timburks commented:
| Are you saying this is a new issue? Or just an observation you happened to perform on this PR?

I believe this was #795. If #796 is merged into this PR (it's in main), I think it will be fixed.

theganyo commented:
> I believe this was #795. If #796 is merged into this PR (it's in main), I think it will be fixed.

done

theganyo commented:
Per offline meeting with @shrutiparabgoogle, controller will continue to support existing syntax for Pattern and `$resource.spec`, no wildcard revisions or revision resource references. However, the generated `controller compute` actions must include the revision in `$resource.spec`. From that, the following are TODOs:

- [ ] Ensure manifest pattern cannot include revisions
- [ ] Ensure generated `controller compute` actions include the revision.

shrutiparabgoogle commented on cmd/registry/controller/controller.go:
```diff
@@ -195,9 +195,21 @@ func generateUpdateActions(
+				pn = r
+			}
+		}
+		visited[pn.String()] = true
```
I think what we need here is ignoring the revision in the parent name while generating the filter (since the filter applies only on the resource name). 

So instead of this, we can make the change in `excludeVisitedParents` like this:
```
func excludeVisitedParents(v map[string]bool) string {
	// Wrap each string with quotes and join them with commas to build a JSON string array.
        // Ignore revision names if any exist in the parent names.
	jsonStrings := make([]string, 0, len(v))
	for parent := range v {
		jsonStrings = append(jsonStrings, fmt.Sprintf("%q", strings.Split(parent, "@")[0]))
	}
	return fmt.Sprintf("[%s].all(parent, !(name==parent))", strings.Join(jsonStrings, ","))
}
```

Thoughts?

theganyo replied:
Yes, that will work - as long as we're ok dropping the guard to check for revision in the pattern, which we don't need now anyway.

shrutiparabgoogle replied:
There is a function to validate the manifest definition: https://github.com/apigee/registry/blob/main/cmd/registry/controller/parser.go#L38

`ParseResource()` doesn't support `SpecRevisions` right now, so that guard is still in place, if we decide to add the specrevision support later, we can modify the validate function accordingly.

shrutiparabgoogle commented on cmd/registry/controller/controller_test.go:
```diff
@@ -774,6 +778,31 @@ func TestMultipleEntitiesArtifacts(t *testing.T) {
 			}
 			actions := ProcessManifest(ctx, lister, projectID, manifest, 10)
 
+			if test.generatedResourceRequiresRevision {
```
I am not sure if we need this flag, wouldn't all generatedResource should include revision IDs in them? Since we are saying that we want to carry through a spec revision till the action.

IMO every reference of a spec should have a revision ID to it, we can't rely on `projects/demo/locations/apis/petstore/versions/v1/specs/openapi.yaml` anywhere because the latest revision might change anytime while the controller is processing the sate and taking actions. Every spec should include the revision ID in the controller.

shrutiparabgoogle commented on cmd/registry/controller/controller_test.go:
```diff
@@ -656,7 +659,7 @@ func TestMultipleEntitiesArtifacts(t *testing.T) {
+			desc: "outdated lastest spec revision artifacts",
 			seed: []seeder.RegistryResource{
 				&rpc.Artifact{
 					Name: "projects/controller-test/locations/global/apis/petstore/versions/1.0.0/specs/openapi.yaml/artifacts/conformance-registry-styleguide",
```
These spec level artifacts should also include revisionID in them.

shrutiparabgoogle commented on cmd/registry/controller/list.go:
```diff
@@ -14,6 +14,7 @@ type listingClient interface {
 	ListAPIs(context.Context, names.Api, string, core.ApiHandler) error
 	ListVersions(context.Context, names.Version, string, core.VersionHandler) error
 	ListSpecs(context.Context, names.Spec, string, core.SpecHandler) error
+	ListSpecRevisions(context.Context, names.SpecRevision, string, core.SpecHandler) error
```
Are we using the ListSpecRevisions anywhere? I think Revision support should be implicitly added to the existing Specs support that we have in the controller.

shrutiparabgoogle commented on cmd/registry/controller/parser.go:
```diff
@@ -61,6 +61,8 @@ func validateGeneratedResourceEntry(parent string, generatedResource *rpc.Genera
 			return resourceName.Version() != ""
 		case "spec":
 			return resourceName.Spec() != ""
+		case "specrev":
```
We are not supporting `$resource.specrev` for now. 

I think if we want to support revisions in some way, we can maybe support `$resource.spec@latest` instead of `specrev`, because which revision to pick will again be a question. This can be a discussion outside the scope of this PR though.

shrutiparabgoogle commented on cmd/registry/controller/parser.go:
```diff
@@ -169,7 +171,10 @@ func generateCommand(action string, resourceName string) (string, error) {
 			entityVal = resource.Version()
 		case "spec":
-			entityVal = resource.Spec()
+			entityVal = resource.SpecRevision()
```
I thought we were not going to add a `SpecRevision` type and going to stick to only `Specs`

theganyo replied:
Sorry, yes, this was done before we discussed not supporting SpecRevisions in the controller... which now it looks like we're going to discuss next week? I imagine I should leave this alone until then.

shrutiparabgoogle commented on cmd/registry/patterns/resources.go:
```diff
@@ -82,6 +87,52 @@ func (s SpecName) ParentName() ResourceName {
 	return VersionName{}
 }
 
+type SpecRevisionName struct {
```
Let's not support both Spec and SpecRevision, that will create a lot of confusion.

shrutiparabgoogle reviewed (commented):
I have reservations on how the controller modifications are made right now, let's sync on this. I will also add some comments on the issue to explain my thought process more clearly.

theganyo commented:
Sorry, yes, there is cleanup on work that was done before we discussed not supporting SpecRevisions in the controller... which now it looks like we're going to discuss next week? I imagine I should leave this alone until then.

theganyo commented:
I removed a bunch of the old stuff related to specrev on the client (cmd/*) and addressed much of the issues mentioned. Not all issues are resolved yet, though, as some commands are failing (see failing tests) because of Spec / SpecRevision naming complexity.

theganyo commented on cmd/registry/patterns/resources.go:
```diff
@@ -41,15 +41,16 @@ type ResourceName interface {
 
 func (s SpecName) Spec() string {
-	return s.Name.String()
+	return s.String()
```
Yes, it does:  https://github.com/apigee/registry/blob/a171157aca4f0d47d09486c97d2c772ee7fa97a2/cmd/registry/patterns/resources.go#L68-L74

theganyo commented on cmd/registry/patterns/parser.go:
```diff
@@ -49,8 +49,8 @@ func parseResource(resourcePattern string) (ResourceName, error) {
 		return VersionName{Name: version}, nil
-	} else if spec, err := names.ParseSpec(resourcePattern); err == nil {
-		return SpecName{Name: spec}, nil
+	} else if spec, err := names.ParseSpecRevision(resourcePattern); err == nil {
```
Just using `names.ParseSpecRevision` should always work as I made the revision part optional. I think your other questions are addressed here:

https://github.com/apigee/registry/blob/a171157aca4f0d47d09486c97d2c772ee7fa97a2/server/registry/actions_spec_revisions_test.go#L330

https://github.com/apigee/registry/blob/a171157aca4f0d47d09486c97d2c772ee7fa97a2/server/registry/actions_spec_revisions_test.go#L300

shrutiparabgoogle replied:
Thanks, this makes sense.

theganyo commented on cmd/registry/patterns/resources.go:
```diff
@@ -213,17 +218,18 @@ func (ar ArtifactName) Artifact() string {
+	if spec, err := names.ParseSpecRevision(specPattern.String()); err == nil {
 		return spec.String()
-	} else if _, err := names.ParseSpecCollection(specPattern.String()); err == nil {
+	} else if _, err := names.ParseSpecRevisionCollection(specPattern.String()); err == nil {
```
Hmm. This looks like a bug that's not being hit in testing. I'll take a look.

theganyo replied:
Since this is preexisting and unclear, I think this deserves a look independent of this PR, I'll create a new issue.

theganyo commented:
@shrutiparabgoogle I updated the [complexity.go](https://github.com/apigee/registry/pull/794/files#diff-a0e166107d9b4f3bb4d456333ba8f8b657bdd8d755485ba18061738bd768a4b0) cmd to work with specs and/or revisions. I assume I'll need to do the same with each of the following. True? Are there any others I'm missing?

- [ ] conformance
- [ ] lint
- [ ] lintStats
- [ ] references
- [ ] score
- [ ] scoreCard
- [ ] vocabulary

shrutiparabgoogle commented:
> @shrutiparabgoogle I updated the [complexity.go](https://github.com/apigee/registry/pull/794/files#diff-a0e166107d9b4f3bb4d456333ba8f8b657bdd8d755485ba18061738bd768a4b0) cmd to work with specs and/or revisions. I assume I'll need to follow that same pattern with each of the following. True? Are there any others I'm missing?
> 
> * [ ]  conformance
> * [ ]  lint
> * [ ]  lintStats
> * [ ]  references
> * [ ]  score
> * [ ]  scoreCard
> * [ ]  vocabulary

`lint` and `lintstats` should be deprecated soon since we have now replaced those with conformance and score commands. Everything else looks correct

theganyo commented:
The latest 2 commits should complete the controller support (aside from lint and lintStats, as you suggested). If you see anything I missed, please let me know.

shrutiparabgoogle commented:
In my testing, I noticed that the conformance command was failing because it needs an update here: https://github.com/apigee/registry/blob/main/cmd/registry/conformance/conformance-task.go#L117

Update to `ParseSpecRevision()`

theganyo commented:
> In my testing, I noticed that the conformance command was failing because it needs an update here: https://github.com/apigee/registry/blob/main/cmd/registry/conformance/conformance-task.go#L117
> 
> Update to `ParseSpecRevision()`

done

shrutiparabgoogle reviewed (approved):
Thanks for adding those parser tests, that gives me confidence that supporting revisions in controller manifest should be straightforward. 

This PR has LGTM from my side, after testing and verifying  the controller changes.

timburks commented:
One observation, which I'm not sure is new: when I list specs with `registry list`, they are in ascending order alphabetically, but when I list revisions, they seem to be in descending order, but perhaps first sorted by revision time and then alphabetically (both descending).

If possible, I think it would be best to sort alphabetically (ascending) by resource name and secondarily sort descending by revision time.
```
$ registry list apis/-/versions/-/specs/-
projects/disco/locations/global/apis/abusiveexperiencereport/versions/v1/specs/discovery.json
projects/disco/locations/global/apis/acceleratedmobilepageurl/versions/v1/specs/discovery.json
projects/disco/locations/global/apis/accessapproval/versions/v1/specs/discovery.json
projects/disco/locations/global/apis/accesscontextmanager/versions/v1/specs/discovery.json
...
```
$ registry list apis/-/versions/-/specs/-@-
projects/disco/locations/global/apis/youtube/versions/v3/specs/discovery.json@258ade1c
projects/disco/locations/global/apis/workflows/versions/v1/specs/discovery.json@16f9d38d
projects/disco/locations/global/apis/workflows/versions/v1beta/specs/discovery.json@d5ffe22e
projects/disco/locations/global/apis/versionhistory/versions/v1/specs/discovery.json@98aa8265
projects/disco/locations/global/apis/transcoder/versions/v1/specs/discovery.json@14f704d2
```

theganyo commented:
Added the existing data migration for spec revision artifacts

timburks reviewed (approved):
LGTM. Apart from tripping over unrelated FK issues, the only concern that I saw while testing this was list order. I mainly exercised "registry compute complexity" and that worked as I expected for both "specs/-" and "specs/-@-". So I think we can go ahead with this, and adding support for deployments can come after this is in.
